cmd/abapEnvironmentAssemblePackages.go

```diff
@@ -0,0 +1,638 @@
+		Username: connectionDetails.User,
+		Password: connectionDetails.Password,
+	})
+	cookieJar, _ := cookiejar.New(nil)
+	conn.Client.SetOptions(piperhttp.ClientOptions{
+		Username:  connectionDetails.User,
+		Password:  connectionDetails.Password,
+		CookieJar: cookieJar,
+	})
+	conn.Baseurl = connectionDetails.URL
+	return nil
+}
+
+// ******** technical communication calls ********
+
+func (conn *connector) getToken() error {
+	conn.Header["X-CSRF-Token"] = []string{"Fetch"}
+	response, err := conn.Client.SendRequest("HEAD", conn.Baseurl, nil, conn.Header, nil)
+	if err != nil {
+		if response == nil {
+			return errors.Wrap(err, "Fetching X-CSRF-Token failed")
+		}
+		defer response.Body.Close()
+		errorbody, _ := ioutil.ReadAll(response.Body)
+		return errors.Wrapf(err, "Fetching X-CSRF-Token failed: %v", string(errorbody))
+
+	}
+	defer response.Body.Close()
+	token := response.Header.Get("X-CSRF-Token")
+	conn.Header["X-CSRF-Token"] = []string{token}
+	return nil
+}
+
+func (conn connector) get(appendum string) ([]byte, error) {
+	url := conn.Baseurl + appendum
+	response, err := conn.Client.SendRequest("GET", url, nil, conn.Header, nil)
+	if err != nil {
+		if response == nil {
+			return nil, errors.Wrap(err, "Get failed")
+		}
+		defer response.Body.Close()
+		errorbody, _ := ioutil.ReadAll(response.Body)
+		return errorbody, errors.Wrapf(err, "Get failed: %v", string(errorbody))
+
+	}
+	defer response.Body.Close()
+	body, err := ioutil.ReadAll(response.Body)
+	return body, err
+}
+
+func (conn connector) post(importBody string) ([]byte, error) {
+	url := conn.Baseurl + "/builds"
+	response, err := conn.Client.SendRequest("POST", url, bytes.NewBuffer([]byte(importBody)), conn.Header, nil)
+	if err != nil {
+		if response == nil {
+			return nil, errors.Wrap(err, "Post failed")
+		}
+		defer response.Body.Close()
+		errorbody, _ := ioutil.ReadAll(response.Body)
+		return errorbody, errors.Wrapf(err, "Post failed: %v", string(errorbody))
+
+	}
+	defer response.Body.Close()
+	body, err := ioutil.ReadAll(response.Body)
+	return body, err
+}
+
+func (conn connector) download(appendum string, downloadPath string) error {
+	url := conn.Baseurl + appendum
+	err := conn.DownloadClient.DownloadFile(url, downloadPath, nil, nil)
+	return err
+}
+
+// ******** BUILD logic ********
+
+func (b *build) start(phase string, inputValues values) error {
+	if err := b.getToken(); err != nil {
+		return err
+	}
+	importBody := inputForPost{
+		phase:  phase,
+		values: inputValues,
+	}.String()
+
+	body, err := b.connector.post(importBody)
+	if err != nil {
+		return err
+	}
+
+	var jBuild jsonBuild
+	json.Unmarshal(body, &jBuild)
+	b.BuildID = jBuild.Build.BuildID
+	b.RunState = jBuild.Build.RunState
+	b.ResultState = jBuild.Build.ResultState
+	b.Phase = jBuild.Build.Phase
+	b.Entitytype = jBuild.Build.Entitytype
+	b.Startedby = jBuild.Build.Startedby
+	b.StartedAt = jBuild.Build.StartedAt
+	b.FinishedAt = jBuild.Build.FinishedAt
```

Comment:
What is this part doing if nothing gets returned?

Reply:
What do yo mean by nothing? If something went wrong the 'b.connector.post(importBody)' part will thrown an error and it returns(therefore never reaches the mentioned code). But if the post went fine, it will return this json-structur(however it will be just partly filled, for example jBuild.Build.FinishedAt will be empty at this point in time almost always.